testsuite/src/main/scala/endpoints/testsuite/BasicAuthTestApi.scala

```diff
@@ -12,4 +13,13 @@ trait BasicAuthTestApi extends algebra.Endpoints with algebra.BasicAuthenticatio
     textResponse
   )
 
+  val protectedEndpointViaBuilder =
+    anEndpoint
+      .withMethod(Get)
+      .withUrl(path / "users")
+      .withTextResponse
+      .withBasicAuth
```

Comment:
Yes, this may cause a lot of trouble when used wrongly.

One solution I see is to make something like this:
```
  case class EndpointBuilder2[Resp, RespMod[_]] {
    def build: Endpoint[???, RespMod[Resp]] = ???
  }
  ```
but this is completely not scalable, because to implement such modifications on all of the elements the number of type arguments grows from 4 to 8(and if we introduce response headers from 5 to 10).

Another approach would be to keep result types in `EndpointBuilder` from the beginning(as type  arguments or type members), like so:
```
  case class EndpointBuilder2[U,RH,RE,Resp, FinalRequest, FinalResponse] {
    def build: Endpoint[FinalRequest, FinalResponse] = ???
  }
```
and the `withResponse` will be awailable when `Resp =:= FinalResponse`

But both of these methods feel more like a hack and I'm not even sure if they will fix anything.